File: common/src/main/java/it/gabliz/additional/common/helper/ModObjectInfo.java
package it.gabliz.additional.common.helper;

import org.jetbrains.annotations.NotNull;

import javax.annotation.Nullable;

public class ModObjectInfo {

    public static class ItemInfo {

        public enum Category  { CRAFTING_ITEM, SEED }

        @NotNull String idName;
        @NotNull String name;
        @NotNull Category category;
        @Nullable String description;
        @Nullable String shiftTooltip;

        public ItemInfo(@NotNull String idName, @NotNull String name, @NotNull Category category) {
            this.idName = idName;
            this.name = name;
            this.category = category;
        }

        public String getIdName() {
            return idName;
        }

        public void setIdName(String idName) {
            this.idName = idName;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public Category getCategory() {
            return category;
        }

        public void setCategory(Category category) {
            this.category = category;
        }

        @Nullable
        public String getDescription() {
            return description;
        }

        public void setDescription(@Nullable String description) {
            this.description = description;
        }

        @Nullable
        public String getShiftTooltip() {
            return shiftTooltip;
        }

        public void setShiftTooltip(@Nullable String shiftTooltip) {
            this.shiftTooltip = shiftTooltip;
        }
    }
}

File: common/src/main/kotlin/it/gabliz/additional/common/helper/ModObjectInfo.kt
package it.gabliz.additional.common.helper

import net.minecraft.data.client.BlockStateModelGenerator
import net.minecraft.data.client.BlockStateSupplier
import net.minecraft.data.client.BlockStateVariantMap
import net.minecraft.item.ItemGroup
import net.minecraft.util.Rarity
import net.minecraft.village.VillagerProfession

interface ModObjectInfo {



    interface Block {

        enum class ModelType { SimpleCubeAll, SimpleBooleanState, CROP_BLOCK }

        val idName : String
        val name : String
        val description : String? get() = null
        val shiftTooltip : String? get() = null
        val model : ModelType get() = ModelType.SimpleCubeAll

        fun getBlockStateMap(blockStateModelGenerator: BlockStateModelGenerator?) : BlockStateVariantMap? = null
        fun getVariantModel(blockStateModelGenerator: BlockStateModelGenerator) : BlockStateSupplier? = null
    }

    interface Item {
        enum class ItemType { CRAFTING_ITEM, SEED }
        val idName : String
        val name : String
        val itemGroup : ItemGroup
        val description : String? get() = null
        val shiftTooltip : String? get() = null
        val hasGlint : Boolean get() = false
        val rarity : Rarity get() = Rarity.COMMON
        val itemType : ItemType get() = ItemType.CRAFTING_ITEM
    }

    interface Villager {
        val poiID : String
        val professionID : String
        val name : String
        val workBlock : net.minecraft.block.Block
        fun registerTrades(profession : VillagerProfession)
    }

}